### PR TITLE
Fixes - Fix publish/draft status for variant content when invariant p…

### LINKED ISSUE #3708: Fix publish/draft status for variant content when invariant properties are edited
child of #3729

Currently if you edit an invariant property (which can only be done on the default language), this does not reflect the publishing state of the default language variant so the user will have no idea if there are actually pending edits.

This problem actually makes it's way into the bulk publishing logic because when invariant properties are edited, this means that the document doesn't show up as a draft and is therefor not published when publishing descendants.

### PR DESCRIPTION
**Pull request by RobertCopilau**
…roperties are edited

### Prerequisites

- [ ] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes: 

### Description

fixes #3708

### CODE REVIEW

File: src/Umbraco.Core/Persistence/Factories/PropertyFactory.cs [hunk 1]
```diff
@@ -103,7 +104,8 @@ public static IEnumerable<PropertyDataDto> BuildDtos(int currentVersionId, int p
             {
                 if (property.PropertyType.IsPublishing)
                 {
-                    var editingCultures = property.PropertyType.VariesByCulture();
+                    var editingCultures = availableCultures?.Any() ?? property.PropertyType.VariesByCulture();
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Core/Persistence/Factories/PropertyFactory.cs]: As far as i can tell we shouldn't be using this `availableCultures` collection here. Firstly we aren't sure that the first one in this list is the 'default' one which is what is being used below. We have an interface to determine the default culture which is: `IDefaultCultureAccessor`. As for the final check below, we should use a check of whether it's an invariant property

File: src/Umbraco.Core/Persistence/Factories/PropertyFactory.cs [hunk 2]
```diff
@@ -131,6 +133,15 @@ public static IEnumerable<PropertyDataDto> BuildDtos(int currentVersionId, int p
                         {
                             editedCultures.Add(propertyValue.Culture); // report culture as edited
                         }
+
+                        // flag culture as edited if it contains an edited invariant property 
+                        if (propertyValue.Culture == null &&
```
Comment (Shazwazza) [inline src/Umbraco.Core/Persistence/Factories/PropertyFactory.cs]: each of these lines should have comments as to what they do just like the code block above since all this code is quite hard to follow

File: src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs [hunk 3]
```diff
@@ -527,7 +527,8 @@ protected override void PersistUpdatedItem(IContent entity)
             Database.Execute(deletePropertyDataSql);
 
             // insert property data
-            var propertyDataDtos = PropertyFactory.BuildDtos(content.VersionId, publishing ? content.PublishedVersionId : 0, entity.Properties, LanguageRepository, out var edited, out var editedCultures);
+            var propertyDataDtos = PropertyFactory.BuildDtos(content.VersionId, publishing ? content.PublishedVersionId : 0,
+                entity.Properties, LanguageRepository, out var edited, out var editedCultures, content.AvailableCultures);
```
Comment (Shazwazza) [inline src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs]: See above comment, i don't think AvailableCultures should be used

**Comment by Shazwazza** [comment]
I've pushed some changes now. My comment about IDefaultCultureAccessor is incorrect since that is only available in the Web project but this method already has access to the ILanguageRepository which can resolve the default language.

I've added comments to this method to explain a bit more. The change I made is to just pass in the `ContentVariations` of the entity, we then just use that to determine if cultures can be edited, we don't need to check the property types variations for that. I also changed some variable names to be less confusing.  I also added a unit test and verified that this test fails without these changes.

I'll merge now but @RobertCopilau please review what I did to confirm I didn't overlook something

**Comment by RobertCopilau** [comment]
Hey Shannon, everything seems in order and it works, thanks for the feedback!
Sometimes I forget that people might not be able to read my code, I will write comments more often.
